elastic/kibana
# [SIEM] Import timeline fix

## Summary

<img width="1674" alt="Screenshot 2020-06-18 at 09 14 13" src="https://user-images.githubusercontent.com/6295984/84995614-21222600-b144-11ea-87a8-80993d1f0ccc.png">


This PR is to allow template timeline updated via import, 
and fix the case we had in https://github.com/elastic/kibana/pull/64439

How to verify this PR:
1. copy the json below and paste it onto a file


```
{"savedObjectId":null,"version":"WzM5LDFd","timelineType":"template","templateTimelineId":"49188240-6530-11ea-90dd-7d87cXavier","templateTimelineVersion":4,"columns":[{"indexes":null,"name":null,"columnHeaderType":"not-filtered","id":"@timestamp","searchable":null},{"indexes":null,"name":null,"columnHeaderType":"not-filtered","id":"message","searchable":null},{"indexes":null,"name":null,"columnHeaderType":"not-filtered","id":"event.category","searchable":null},{"indexes":null,"name":null,"columnHeaderType":"not-filtered","id":"event.action","searchable":null},{"indexes":null,"name":null,"columnHeaderType":"not-filtered","id":"host.name","searchable":null},{"indexes":null,"name":null,"columnHeaderType":"not-filtered","id":"source.ip","searchable":null},{"indexes":null,"name":null,"columnHeaderType":"not-filtered","id":"destination.ip","searchable":null},{"indexes":null,"name":null,"columnHeaderType":"not-filtered","id":"user.name","searchable":null}],"dataProviders":[],"description":"is Super","eventType":"all","filters":[],"kqlMode":"filter","kqlQuery":{"filterQuery":{"serializedQuery":"{\"bool\":{\"should\":[{\"exists\":{\"field\":\"host.name\"}}],\"minimum_should_match\":1}}","kuery":{"expression":"host.name: *","kind":"kuery"}}},"title":"My template timeline updated","dateRange":{"start":1584020448645,"end":1584106848645},"savedQueryId":null,"sort":{"columnId":"@timestamp","sortDirection":"desc"},"favorite":[{"favoriteDate":1584480703444,"keySearch":"WGF2aWVyTQ==","fullName":"xavier mouligneau","userName":"XavierM"}],"created":1588693289719,"createdBy":"XavierM","updated":1588693315060,"updatedBy":"XavierM","eventNotes":[{"noteId":"e3cf0900-8ee6-11ea-bff9-5be90e832634","version":"WzM0LDFd","eventId":"Cfwg1HABO74GafNORxyF","note":"pinned and event note","timelineId":"e37d8da0-8ee6-11ea-bff9-5be90e832634","created":1588693290295,"createdBy":"XavierM","updated":1588693290295,"updatedBy":"XavierM"},{"noteId":"e3cfcc50-8ee6-11ea-bff9-5be90e832634","version":"WzM1LDFd","eventId":"Cvwg1HABO74GafNORxyF","note":"event note 1","timelineId":"e37d8da0-8ee6-11ea-bff9-5be90e832634","created":1588693290295,"createdBy":"XavierM","updated":1588693290295,"updatedBy":"XavierM"}],"globalNotes":[{"noteId":"e3cebae0-8ee6-11ea-bff9-5be90e832634","version":"WzI2LDFd","note":"Global","timelineId":"e37d8da0-8ee6-11ea-bff9-5be90e832634","created":1588693290295,"createdBy":"XavierM","updated":1588693290295,"updatedBy":"XavierM"},{"noteId":"e3cf3010-8ee6-11ea-bff9-5be90e832634","version":"WzM2LDFd","note":"Note","timelineId":"e37d8da0-8ee6-11ea-bff9-5be90e832634","created":1588693290295,"createdBy":"XavierM","updated":1588693290295,"updatedBy":"XavierM"}],"pinnedEventIds":["Yfwg1HABO74GafNObyFv","Cfwg1HABO74GafNORxyF","Cvwg1HABO74GafNORxyF"]}
```
2. save it as template_timeline.ndjson
3. import in to timeline
4. More tweak: https://github.com/elastic/kibana/pull/65448#issuecomment-636748734

Error cases:

 1. Import a new timeline (Create timeline via import)
	- input validation:
		- `timelineId` has to be null / empty or a timelineId that doesn't exist in SO - if given a timelineId that's already exist,  therefor throwing error: ***savedObjectId: "${id}" already exists***
		- `title` must have - if not given, throw error: ***Title cannot be empty***
		- version: has to be null (this is to reserve the possibility of version control for updating timeline in the future)
		- `timelineType` has to be null / default or not specified - if set to template, I'll assume it is trying to create a template
		- `status` has to be null / 'active' or not specified - we don't expose this field to exported file. So far we do not block the 'immutable' status as we need that when installing Elastic template. If `status` is given as **draft**, throw error ***Cannot create a draft timeline***
 2. Import an existing timeline (Update existing timeline via import)
 	- We don't allow updating existing timeline atm: therefor throwing error: ***savedObjectId: "${id}" already exists***

 3. Import a new template (Create custom template via import)
	- input validation:
		- `templateTimelineId` has to be given a uuid that doesn't exist in SO, if given a timelineId that's already exist,  therefor throwing error: ***savedObjectId: "${id}" already exists***; if not given a templateTimelineId, throw error: ***Create template timeline without a template timeline ID is not allowed***
		- `title` must have - if not given, throw error: ***Title cannot be empty***
		- `timelineType` has to be template - if set to others, I'll assume it is trying to create a default timeline
		- `status` has to be null / 'active' or not specified - we don't expose this field to exported file. So far we do not block the 'immutable' status as we need that when installing Elastic template. If `status` is given as **draft**, throw error ***Cannot create a draft timeline***
 4. Import an existing template
 	- input validation:
		- `timelineType` has to be the same as existing template timeline, otherwise throw error: ***'Update timelineType is not allowed'***
		- `status` has to be the same as existing template timeline. Otherwise throw error ***Update status is not allowed***
		- `templateTimelineId` has to be given a uuid that matches an exist in SO. 	
			- If not given: ***template timeline without a template timeline ID is not allowed***
			-  ~~if given a timelineId that's not exist,  therefor throwing error: ***CREATE template timeline with PATCH is not allowed, please use POST instead (Given template timeline doesn't exist)***~~
			- given timelineId doesn't match: ***Timeline id doesn't match with existing template timeline***
			- given templateTimelineVersion doesn't match: ***TimelineVersion conflict: The given version doesn not match with existing timeline***

### Checklist

Delete any items that are not applicable to this PR.


### For maintainers
